server/src/main/resources/org/elasticsearch/bootstrap/test-framework.policy

```diff
@@ -125,4 +125,10 @@ grant codeBase "${codebase.netty-transport}" {
 grant {
   permission java.net.SocketPermission "127.0.0.1", "connect, resolve";
   permission java.nio.file.LinkPermission "symbolic";
+
+  // needed for com.nimbusds.jose.shaded.gson.Gson.toJson
+  permission java.lang.RuntimePermission "accessDeclaredMembers";
+
+  // needed for com.nimbusds.oauth2.sdk.id.Identifier#value
+  permission java.lang.reflect.ReflectPermission "suppressAccessChecks";
```

Comment:
Can we grant these permissions be granted to the nimbusds codebase, rather than the "global" grant?

Reply:
I agree it would be better to scope it to the two jars. However, I tried this and it didn't work.

```
grant codeBase "${codebase.nimbus-jose-jwt}" {
  // needed for com.nimbusds.jose.shaded.gson.Gson.toJson
  permission java.lang.RuntimePermission "accessDeclaredMembers";
};

grant codeBase "${codebase.oauth2-oidc-sdk}" {
  // needed for com.nimbusds.oauth2.sdk.id.Identifier#value
  permission java.lang.reflect.ReflectPermission "suppressAccessChecks";
};
```

I tried adding `nimbus-jose-jwt` and `oauth2-oidc-sdk` entries here, but it didn't work.
https://github.com/elastic/elasticsearch/blob/64728c08422ce2d493502809bb09979a4aa6852b/test/framework/src/main/java/org/elasticsearch/bootstrap/BootstrapForTesting.java#L212-L224

There is a check here which rejects classes in third-party jar files, so it was not clear to me how to add the two codebases without interfering with the existing, internal codebases.
https://github.com/elastic/elasticsearch/blob/64728c08422ce2d493502809bb09979a4aa6852b/test/framework/src/main/java/org/elasticsearch/bootstrap/BootstrapForTesting.java#L227-L243

Reply:
This is what I tried to do:
```
addClassCodebase(codebases, "nimbus-jose-jwt", "com.nimbusds.jose.shaded.gson.Gson.toJson");
addClassCodebase(codebases, "oauth2-oidc-sdk", "com.nimbusds.oauth2.sdk.id.Identifier");
```

The `location` value in `addClassCodebase` ends with `.jar`, so this code falls through without registering the codebase.
```
if (location.toString().endsWith(".jar") == false) {
```

References to `codebase.nimbus-jose-jwt` and `codebase.oauth2-oidc-sdk` inside the policy files are unresolved, so the policy grant cannot be parsed correctly.

Reply:
@ChrisHegarty, I tried moving the policy changes from `server/src/main/resources/org/elasticsearch/bootstrap/security.policy` to `x-pack/plugin/security/src/main/plugin-metadata/plugin-security.policy`, but this part of the build failed: `./gradlew :x-pack:qa:oidc-op-tests:javaRestTest --debug --stacktrace`

Success (server): https://gradle-enterprise.elastic.co/s/mr3onfehxffj2/console-log#L886
Failure (x-pack security): https://gradle-enterprise.elastic.co/s/qdi5gmcqzq65i/console-log#L748

Does the same permission need to be added elsewhere? I looked under /x-pack/security/qa/, but I don't see a policy file in there.

Reply:
I debugged PolicyUtil and fo0und the grant names for these jars include the version and jar, so I updated the grants. I double checked and those policies work in unit tests. However, they don't help inside the elasticsearch:test container of the test fixture. In other words, SecurityManager policy seems to be treated differently inside vs outside the container.